Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py

```diff
@@ -308,6 +308,63 @@ def to_identifier(self, hash=None, revision=None, populate=True, branch=None):
                     return self.to_identifier(hash=hash, populate=False)
             return None
 
+    class BranchEnvironment(object):
+        BASE = 'environment'
+
+        def __init__(self, repository):
+            self._repository = repository
+
+        def cleanup(self):
+            branches = self._repository.branches
+            for line in run(
+                [self._repository.executable(), 'config', '--get-regexp', self.BASE],
+                capture_output=True,
+                cwd=self._repository.path,
+                encoding='utf-8',
+            ).stdout.splitlines():
+                key, _ = line.split(maxsplit=1)
+                components = key.split('.')
+                if len(components) != 3:
+                    continue
+                if components[1] in branches:
+                    continue
+                if run(
+                    [self._repository.executable(), 'config', '--unset', key],
+                    capture_output=True,
+                    cwd=self._repository.path,
+                    encoding='utf-8',
+                ).returncode:
+                    sys.stderr.write("Failed to clean up environment for '{}'\n".format(components[1]))
+
+        def values(self, branch=None):
+            result = {}
+            for line in run(
+                [self._repository.executable(), 'config', '--get-regexp', self.BASE],
+                capture_output=True,
+                cwd=self._repository.path,
+                encoding='utf-8',
+            ).stdout.splitlines():
+                full_key, value = line.split(maxsplit=1)
```

Comment:
Do we really not have any generic code to do this? Because this (and in `prepare-commit-msg` above) does seem somewhat fragile—they rely on the values to not contain new lines (c.f. `--null`) and for the config file to be valid utf-8 (which, to be fair, it probably will be… but git doesn't actually guarantee/enforce that).

Not gonna say we need to fix this before we land _this_, but it would be good to add something before this gets worse.

Reply:
Not ignoring this feedback, but the change I just uploaded essentially makes it obsolete.

We do have generic code to do this, but I was trying to limit the scope of `config` values we were extracting. Also, having a hook share code with the repository it's operating on is somewhat unreliable, so I try to avoid it where possible, hence the duplication in that change. New change uses a different approach, so there won't be duplication.